src/invert/laplace/impls/iterative_parallel_tri/iterative_parallel_tri.cxx

```diff
@@ -0,0 +1,1597 @@
+    if ((global_flags & INVERT_KX_ZERO) && (kz == 0)) {
+      dcomplex offset(0.0);
+      for (int ix = localmesh->xstart; ix <= localmesh->xend; ix++) {
+	offset += xk1d(kz,ix);
+      }
+      offset /= static_cast<BoutReal>(localmesh->xend - localmesh->xstart + 1);
+      for (int ix = localmesh->xstart; ix <= localmesh->xend; ix++) {
+	xk1d(kz,ix) -= offset;
+      }
+    }
+
+    // Store the solution xk for the current fourier mode in a 2D array
+    for (int ix = 0; ix < ncx; ix++) {
+      xk(ix, kz) = xk1d(kz,ix);
+    }
+    first_call(jy,kz) = false;
+  }
+  ///SCOREP_USER_REGION_END(afterloop);
+
+  ///SCOREP_USER_REGION_DEFINE(fftback);
+  ///SCOREP_USER_REGION_BEGIN(fftback, "fft back",///SCOREP_USER_REGION_TYPE_COMMON);
+  // Done inversion, transform back
+  for (int ix = 0; ix < ncx; ix++) {
+
+    if(global_flags & INVERT_ZERO_DC)
+      xk(ix, 0) = 0.0;
+
+    irfft(&xk(ix, 0), ncz, x[ix]);
+
+#if CHECK > 2
+    for(int kz=0;kz<ncz;kz++)
+      if(!finite(x(ix,kz)))
+	throw BoutException("Non-finite at %d, %d, %d", ix, jy, kz);
+#endif
+  }
+  ///SCOREP_USER_REGION_END(fftback);
+  return x; // Result of the inversion
+}
+
+/*
+ * Perform Gauss-Seidel with red-black colouring on the reduced system. 
+ * We don't attempt comm/comp overlap, as there is not sigificant work in the
+ * x loop.
+ */
+void LaplaceIPT::gauss_seidel_red_black(Level &l, const Array<bool> &converged, const int jy){
+
+  SCOREP0();
+
+  if(l.included){
+
+    Array<dcomplex> sendvec, recvecin, recvecout;
+    sendvec = Array<dcomplex>(nmode);
+    recvecin = Array<dcomplex>(nmode);
+    recvecout = Array<dcomplex>(nmode);
+    MPI_Request rreqin, rreqout, sreqin, sreqout;
+
+    // BLACK SWEEP
+    // 
+    // Red processors communication only
+    if(l.red){
+      // Post receives
+      if(not localmesh->firstX()){
+	MPI_Irecv(&recvecin[0], nmode, MPI_DOUBLE_COMPLEX, l.proc_in, 0, BoutComm::get(), &rreqin);
+      }
+      if(not localmesh->lastX()){ // this is always be true is we force an even core count
+	MPI_Irecv(&recvecout[0], nmode, MPI_DOUBLE_COMPLEX, l.proc_out, 1, BoutComm::get(), &rreqout);
+      }
+
+      // Receive and put data in arrays
+      if(!localmesh->firstX()){
+	MPI_Wait(&rreqin,MPI_STATUS_IGNORE);
+	for(int kz=0; kz < nmode; kz++){
+	  if(!converged[kz]){
+	    l.xloc(0,kz) = recvecin[kz];
+	  }
+	}
+      }
+      if(!localmesh->lastX()){
+	MPI_Wait(&rreqout,MPI_STATUS_IGNORE);
+	for(int kz=0; kz < nmode; kz++){
+	  if(!converged[kz]){
+	    l.xloc(3,kz) = recvecout[kz];
+	  }
+	}
+      }
+    }
+
+    // Black processors: work and communication
+    if(l.black){
+      // Black processors do work
+      for (int kz = 0; kz <= maxmode; kz++) {
+	if(!converged[kz]){
+	  // Due to extra point on final proc, indexing of last term is 2, not 3. To remove branching, this is handled by l.index_end
+	  l.xloc(1,kz) = ( l.rr(1,kz) - l.ar(jy,1,kz)*l.xloc(0,kz) - l.cr(jy,1,kz)*l.xloc(l.index_end,kz) ) / l.br(jy,1,kz);
```

Comment:
Precomputing is `1/l.br` is worth doing.

I wondered that too about the `converged` check.  I haven't tried removing it, but I do know something relevant.  At scale, the bottleneck is the MPI_Allreduce in checking whether kz modes are converged.  If you only do the checks infrequently, you save time from not doing the allreduce, but this is offset by an increase in time spent doing calculations for modes that are converged (but you don't know are converged, because you skipped the check).  This says the work* in these loops is worth masking out.

* It is the work that's the cost; we do the communication (but not the buffer packing) for all modes.